src/ui/simulator/toolbox/input/connection.cpp

```diff
@@ -173,6 +173,121 @@ void Connections::internalBuildSubControls()
     sizer->Layout();
 }
 
+static void addUpstreamDownstream(wxTreeItemId& rootId,
+                                  wxTreeCtrl* listbox,
+                                  const Data::AreaList& areas,
+                                  size_t layerID)
+{
+    auto localRootId = listbox->AppendItem(rootId, wxString(wxT("Upstream / Downstream")), 2);
+    listbox->SetItemBold(localRootId, true);
+    {
+        for (const auto& namedArea : areas)
+        {
+            Data::Area* area = namedArea.second;
+            if (area->isVisibleOnLayer(layerID))
+            {
+                wxTreeItemId id;
+                // Foreach Interconnection for the area
+                const Data::AreaLink::Map::iterator end = area->links.end();
+                for (Data::AreaLink::Map::iterator i = area->links.begin(); i != end; ++i)
+                {
+                    Data::AreaLink* lnk = i->second;
+
+                    if (lnk->isVisibleOnLayer(layerID))
+                    {
+                        if (!id)
+                        {
+                            // We have to create the item corresponding to the area
+                            id = listbox->AppendItem(
+                              localRootId,
+                              wxString() << wxT(' ') << wxStringFromUTF8(area->name) << wxT(' '),
+                              1,
+                              1);
+                            listbox->SetItemBold(id, true);
+                        }
+                        // Adding the item for the interconnection
+                        /* wxTreeItemId itemID = */ listbox->AppendItem(
+                          id, /*parent*/
+                          // caption
+                          wxString() << wxT(' ') << wxStringFromUTF8(lnk->with->name) << wxT(' '),
+                          0,
+                          0,
+                          new TreeLeaf(lnk));
+                    }
+                }
+            }
+        }
+    }
+}
+
+static void addByArea(wxTreeItemId& rootId,
+                      wxTreeCtrl* listbox,
+                      const Data::AreaList& areas,
+                      size_t layerID)
+{
+    // Root Node
+    auto localRootId = listbox->AppendItem(rootId, wxString(wxT("By area")), 2);
+    listbox->SetItemBold(localRootId, true);
+
+    // For each area, list the links related to it. For each link, the area is either
+    // its origin or extremity.
+    using ListOfLinks
+      = std::vector<std::pair<Data::AreaLink* /*pointer to link*/,
+                              bool /*true if area is the origin of the link, false otherwise*/>>;
+    // 1. Build hierarchy
+    std::map<Data::AreaName, ListOfLinks> areaToListOfLinks;
```

Comment:
What if, instead of using a `map<AreaName, vector<pair<Data::AreaLink*, bool>>`, we use a `map<AreaName, vector<Data::AreaName>>` ?
So **areaToListOfLinks** would change into **areaToListOfAreas**
If so, I guess it would result in a bit more simple code (see further).

Reply:
Keeping pointers is easier. If we keep names, we need to lookup names in a table.